### ISSUE TITLE
GE2/1 OH promless programming

### ISSUE THREAD

**Issue by mexanick** [evka85/GEM_AMC#29]
## Brief summary of issue

Currently the CTP7 FW doesn't allow to promlessly program the GE2/1 optohybrid without executing special SCA command. As a consequence, the special script has to be used and the procedure is slightly overcompilcated. 
### Types of issue

- [ ] Bug report (report an issue with the code)
- [x] Feature request (request for change which adds functionality)

## Expected Behavior

OH promless programming should be triggered by sending a hard reset from the CTP7 (or TTC stream)
## Current Behavior

OH programming doesn't happen upon receiving a hard reset. 

## Your Environment

* Version used: CTP7 FW release 3.8.0
* Shell used:

**Comment by bdorney** [comment]
Just a quick follow-up question.

I’ve noticed that the OH FW occasionally is “lost” from the Zynq processor RAM and needs to be reloaded. Also I noticed sometimes it takes multiple reloads to “stick.”  

Did you try calling gemloader_configure.sh multiple times?

**Comment by bdorney** [comment]
Occasionally also multiple SCA resets are needed.

**Comment by evka85** [comment]
Hi Brian,

This is actually due to the GE2/1 OH having a different mapping of the SCA GPIO, so the reset procedure is different on GE2/1, and I have to update the CTP7 firmware to execute the reset correctly. The script that I wrote does the reset from software manually controlling the SCA GPIO. Also some GPIOs have to be set to tell the OH to actually enable the programming over GBTX (this will likely be removed in the next version of the OH).

Anyway, this is on my todo list, and I'll implement it soon. I'll reply here once it's done. Right now just running the script gets the job done.

**Comment by bdorney** [comment]
@evka85 thanks for this.  Out of curiosity is this an issue of how the GPIO settings in the SCA are set?  If so couldn't the user just change the GPIO direction and also the GPIO values?  We did this awhile for GE1/1 before it was defaulted in the CTP7 FW (circa 3.5.Y or 3.4.Y?).

```bash
$ sca.py -h
usage: sca.py [-h]
              cardName ohMask
              {compare-mcs-bit,h,hh,fpga-id,program-fpga,gpio-read-input,gpio-set-direction,gpio-set-output,r,sysmon}
              ...

Arguments to supply to sca.py

positional arguments:
  cardName              hostname of the AMC you are connecting too, e.g.
                        'eagle64'; if running on an AMC use 'local' instead
  ohMask                ohMask to apply, a 1 in the n^th bit indicates the
                        n^th OH should be considered
  {compare-mcs-bit,h,hh,fpga-id,program-fpga,gpio-read-input,gpio-set-direction,gpio-set-output,r,sysmon}
                        sca command help
    compare-mcs-bit     compares a *.mcs with a *.bit file to check if the
                        firmware is the same
    h                   FPGA hard reset will be done
    hh                  FPGA hard reset will be asserted and held
    fpga-id             FPGA ID will be read through JTAG
    program-fpga        Program OH FPGA with a bitfile or an MCS file
    gpio-read-input     Read GPIO settings of the SCA
    gpio-set-direction  Set the GPIO Direction Mask
    gpio-set-output     Set the GPIO output values
    r                   SCA reset will be done
    sysmon              Read FPGA sysmon data repeatedly

optional arguments:
  -h, --help            show this help message and exit
```

**Comment by evka85** [comment]
Yes, that's exactly what my script is doing -- it's configuring the GPIO manually, and it works fine.

Here's the code I'm using:
    gpio_dir = 0xff0fe0                                                                    
    gpio_default_out = 0x60                                                                
    gpio_hr_out = 0xff0fe0                                                                 
                                                                                           
    subheading('Setting the GPIO direction mask to ' + hex(gpio_dir))                      
    sendScaCommand(ohList, 0x2, 0x20, 0x4, gpio_dir, False)                                
    sleep(0.1)                                                                             

    subheading('Setting hard reset')                                 
    sendScaCommand(ohList, 0x2, 0x10, 0x4, gpio_hr_out, False)       
    sleep(0.01)                                                      

    subheading('Unsetting hard reset')                                                               
    sendScaCommand(ohList, 0x2, 0x10, 0x4, gpio_default_out, False)                                  
    sleep(0.01)                                                                                      

    subheading('Executing PROMless programming')                                                     
    writeReg(getNode('GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET'), 0x1)                                
    sleep(0.15)                                                                                      

I'm also reading back the FPGA-DONE pin with this:
    readData = sendScaCommand(ohList, 0x2, 0x1, 0x1, 0x0, True)                                      
    subheading('Read GPIO = %s' % hex(readData[0]))                                                  
    fpga_done = (readData[0] >> 30) & 1

**Comment by evka85** [comment]
Well, I mean my script is using the internal functions of the sca.py to do all the things it needs to do. But you're right, even a standard  sca.py script can be used with commands like gpio-set-direction and gpio-set-output to achieve the same result (it would just be a bit slower)
